File: src/Yadasm/Archs/Arch65C816.hs
module Yadasm.Archs.Arch65C816 where

import qualified Yadasm.Node as N
import qualified Yadasm.Binary as B
import qualified Data.ByteString as ByteString
import qualified Yadasm.Symbol as S
import qualified Yadasm.Line as L
import qualified Yadasm.Context as C
import qualified Yadasm.Comparator as Cmp
import           Text.Printf
import           Data.Maybe (isNothing)
import           Yadasm.Archs.Arch6502Helpers
import qualified Yadasm.Archs.Arch6502 as A6502
import qualified Yadasm.Archs.Arch65C02 as A65C02

mask :: InstructionMode -> Integer -> Integer
mask StackS opcode
  | isCC11 opcode = makeOpcode 0x03 opcode
mask DirectPageIndirectLong opcode
  | isCC11 opcode = makeOpcode 0x07 opcode
mask AbsoluteLong opcode
  | isCC11 opcode = makeOpcode 0x0F opcode
mask StackSY opcode
  | isCC11 opcode = makeOpcode 0x13 opcode
mask DirectPageIndirectLongY opcode
  | isCC11 opcode = makeOpcode 0x17 opcode
mask AbsoluteLongX opcode
  | isCC11 opcode = makeOpcode 0x1F opcode
mask im opcode = maskEcho im opcode

makeImInfo :: String -> InstructionMode -> Integer -> ImInfo
makeImInfo = makeImInfo' mask

makeExtended :: String -> Integer -> [ImInfo]
makeExtended name opcode =
  [ makeImInfo name AbsoluteLong opcode
  , makeImInfo name AbsoluteLongX opcode
  , makeImInfo name DirectPageIndirectLong opcode
  , makeImInfo name DirectPageIndirectLongY opcode
  , makeImInfo name StackS opcode
  , makeImInfo name StackSY opcode]

makeCop :: String -> Integer -> [ImInfo]
makeCop name opcode = [makeImInfo name ImmediateByte opcode]

makeJumpLong :: String -> Integer -> [ImInfo]
makeJumpLong name opcode = [makeImInfo name JumpAbsoluteLong opcode]

makePer :: String -> Integer -> [ImInfo]
makePer name opcode = [makeImInfo name Per opcode]

makeBrl :: String -> Integer -> [ImInfo]
makeBrl name opcode = [makeImInfo name BranchLong opcode]

makeMove :: String -> Integer -> [ImInfo]
makeMove name opcode = [makeImInfo name Move opcode]

makePei :: String -> Integer -> [ImInfo]
makePei name opcode = [makeImInfo name ZeroPage opcode]

makePea :: String -> Integer -> [ImInfo]
makePea name opcode = [makeImInfo name Immediate opcode]

makeJml :: String -> Integer -> [ImInfo]
makeJml name opcode = [makeImInfo name Jml opcode]

makeJsrx :: String -> Integer -> [ImInfo]
makeJsrx name opcode = [makeImInfo name Jsrx opcode]

nodes' :: ([ImInfo] -> [N.Node]) -> [N.Node]
nodes'
  makeInstructions = A65C02.nodes' makeInstructions ++ node1 ++ node2 ++ node3
  where
    -- again split up as a workaround to lsp 
    -- having issues with large ++ chains
    node1 = makeInstructions
      (makeExtended "ora" 0x03
       ++ makeExtended "and" 0x23
       ++ makeExtended "eor" 0x43
       ++ makeExtended "adc" 0x63
       ++ makeExtended "sta" 0x83
       ++ makeExtended "lda" 0xA3
       ++ makeExtended "cmp" 0xC3
       ++ makeExtended "sbc" 0xE3
       ++ A6502.makeImplied "xce" 0xFB
       ++ makeCop "cop" 0x02)

    node2 = makeInstructions
      (A6502.makeImplied "phd" 0x0B
       ++ A6502.makeImplied "pld" 0x2B
       ++ A6502.makeImplied "phk" 0x4B
       ++ A6502.makeImplied "rtl" 0x6B
       ++ A6502.makeImplied "phb" 0x8B
       ++ A6502.makeImplied "plb" 0xAB
       ++ A6502.makeImplied "wai" 0xCB
       ++ A6502.makeImplied "xba" 0xEB
       ++ A6502.makeImplied "tcs" 0x1B
       ++ A6502.makeImplied "tsc" 0x3B
       ++ A6502.makeImplied "tcd" 0x5B
       ++ A6502.makeImplied "tdc" 0x7B
       ++ A6502.makeImplied "txy" 0x9B
       ++ A6502.makeImplied "tyx" 0xBB
       ++ A6502.makeImplied "stp" 0xDB
       ++ A6502.makeImplied "wdm" 0x42)

    node3 = makeInstructions
      (makeJumpLong "jsl" 0x22
       ++ makePer "per" 0x62
       ++ makeBrl "brl" 0x82
       ++ makeCop "rep" 0xC2
       ++ makeCop "sep" 0xE2
       ++ makeMove "mvp" 0x44
       ++ makeMove "mvn" 0x54
       ++ makePei "pei" 0xD4
       ++ makePea "pea" 0xF4
       ++ makeJumpLong "jmp" 0x5C
       ++ makeJml "jmp" 0xDC
       ++ makeJsrx "jsr" 0xFC)

nodes :: [N.Node]
nodes = nodes' (makeInstructions' 2)

nodesEmulated :: [N.Node]
nodesEmulated = nodes' (makeInstructions' 1)

File: lib/src/dasm/archs/arch65/arch65c816.rs
use crate::dasm::{Node, Word};

use super::{
    is_cc11, make_implied, make_instructions65c02, make_opcode, mask_echo, ImInfo, InstModes,
};

pub fn mask(im: InstModes, opcode: Word) -> Word {
    match im {
        InstModes::StackS => make_opcode(0x03, opcode),
        InstModes::DirectPageIndirectLong if is_cc11(opcode) => make_opcode(0x07, opcode),
        InstModes::AbsoluteLong if is_cc11(opcode) => make_opcode(0x0F, opcode),
        InstModes::StackSY if is_cc11(opcode) => make_opcode(0x13, opcode),
        InstModes::DirectPageIndirectLongY if is_cc11(opcode) => make_opcode(0x17, opcode),
        InstModes::AbsoluteLongX if is_cc11(opcode) => make_opcode(0x1F, opcode),
        _ => mask_echo(im, opcode),
    }
}

fn make_im_info(im: InstModes, name: &str, opcode: Word) -> ImInfo {
    ImInfo::new(mask, im, name, opcode)
}

fn make_im_from(ims: &[InstModes], name: &str, opcode: Word) -> Vec<ImInfo> {
    ims.iter().map(|i| make_im_info(*i, name, opcode)).collect()
}

fn make_extended(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(
        &[
            InstModes::AbsoluteLong,
            InstModes::AbsoluteLongX,
            InstModes::DirectPageIndirectLong,
            InstModes::DirectPageIndirectLongY,
            InstModes::StackS,
            InstModes::StackSY,
        ],
        name,
        opcode,
    )
}

fn make_cop(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::ImmediateByte], name, opcode)
}

fn make_jump_long(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::JumpAbsoluteLong], name, opcode)
}

fn make_per(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::Per], name, opcode)
}

fn make_brl(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::BranchLong], name, opcode)
}

fn make_move(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::Move], name, opcode)
}

fn make_pei(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::ZeroPage], name, opcode)
}

fn make_pea(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::Immediate], name, opcode)
}

fn make_jml(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::Jml], name, opcode)
}

fn make_jsrx(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::Jsrx], name, opcode)
}

pub fn make_instructions65c816(immediate_size: usize) -> Vec<Node> {
    let mut res = make_instructions65c02(immediate_size);

    [
        make_extended("ora", 0x03),
        make_extended("eor", 0x43),
        make_extended("adc", 0x63),
        make_extended("sta", 0x83),
        make_extended("lda", 0xA3),
        make_extended("cmp", 0xC3),
        make_extended("sbc", 0xE3),
        make_implied("xce", 0xFB),
        make_cop("cop", 0x02),
        make_implied("phd", 0x0B),
        make_implied("pld", 0x2B),
        make_implied("rtl", 0x6B),
        make_implied("phb", 0x8B),
        make_implied("plb", 0xAB),
        make_implied("wai", 0xCB),
        make_implied("xba", 0xEB),
        make_implied("tcs", 0x1B),
        make_implied("tsc", 0x3B),
        make_implied("tcd", 0x5B),
        make_implied("tdc", 0x7B),
        make_implied("txy", 0x9B),
        make_implied("tyx", 0xBB),
        make_implied("stp", 0xDB),
        make_implied("wdm", 0x42),
        make_jump_long("jsl", 0x22),
        make_per("per", 0x62),
        make_brl("brl", 0x82),
        make_cop("rep", 0xC2),
        make_cop("sep", 0xE2),
        make_move("mvp", 0x44),
        make_move("mvn", 0x54),
        make_pei("pei", 0xD4),
        make_pea("pea", 0xF4),
        make_jump_long("jmp", 0x5C),
        make_jml("jmp", 0xDC),
        make_jsrx("jsr", 0xFC),
    ]
    .iter()
    .for_each(|i| {
        res.append(&mut super::arch65helpers::make_instructions(
            i,
            immediate_size,
        ))
    });

    res
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        dasm::archs::{make_arch, IMMEDIATE_SIZE16},
        dasm::{parse_to_strings, Arch, Context, Error},
    };

    fn test_context() -> Context {
        let ctx = Context::new(0x600, 0x700);

        ctx
    }

    fn test_arch() -> Arch {
        make_arch(&make_instructions65c816(IMMEDIATE_SIZE16), None)
    }

    fn parser_helper(data: &[u8], expected: &[&str]) -> (Vec<String>, Result<Vec<String>, Error>) {
        let mut ctx = test_context();
        let arch = test_arch();

        let result = parse_to_strings(&mut ctx, data, &[&arch]);
        let expected = expected.iter().map(|s| s.to_string()).collect();

        (expected, result)
    }

    #[test]
    fn it_should_parse_long_branch_and_immediate() {
        let (expected, result) = parser_helper(
            &[0xA9, 0x12, 0x34, 0x82, 0x6A, 0xFF],
            &["lda #$3412", "brl label_56F"],
        );
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_extended_instructions() {
        let (expected, result) = parser_helper(
            &[
                0x03, 0x11, 0x07, 0x22, 0x0F, 0x33, 0x44, 0x55, 0x13, 0x66, 0x17, 0x77, 0x1F, 0xAA,
                0xBB, 0xCC,
            ],
            &[
                "ora $11, s",
                "ora [$22]",
                "ora $554433",
                "ora ($66, s), y",
                "ora [$77], y",
                "ora $CCBBAA, x",
            ],
        );
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_xce() {
        let (expected, result) = parser_helper(&[0xFB], &["xce"]);
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_cop() {
        let (expected, result) = parser_helper(&[0x02, 0x12], &["cop #$12"]);
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_long_jump() {
        let (expected, result) = parser_helper(
            &[0x22, 0x12, 0x23, 0x45, 0x22, 0x12, 0x23, 0x45],
            &["jsl label_452312", "jsl label_452312"],
        );
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_per() {
        let (expected, result) = parser_helper(&[0x62, 0x12, 0x23], &["per label_2312"]);
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_rep_and_sep() {
        let (expected, result) =
            parser_helper(&[0xC2, 0x12, 0xE2, 0x13], &["rep #$12", "sep #$13"]);
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_move() {
        let (expected, result) = parser_helper(&[0x44, 0x12, 0xE2], &["mvp #$12, #$E2"]);
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_pea() {
        let (expected, result) = parser_helper(&[0xF4, 0x12, 0x11], &["pea #$1112"]);
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_jml() {
        let (expected, result) = parser_helper(&[0xDC, 0x12, 0x11], &["jmp [label_1112]"]);
        assert_eq!(Ok(expected), result);
    }

    #[test]
    fn it_should_parse_jsrx() {
        let (expected, result) = parser_helper(&[0xFC, 0x12, 0x11], &["jsr (label_1112, x)"]);
        assert_eq!(Ok(expected), result);
    }
}
